src/utils/bufferedAsyncIterator.js

```diff
@@ -0,0 +1,94 @@
+const EventEmitter = require('events')
+
+const createPromiseNotifier = (emitter, results, handleError) => (promise, i) => {
+  return promise
+    .then(result => {
+      if (emitter.listenerCount('data') > 0) {
+        emitter.emit('data', { result, id: i })
+      } else {
+        results.push(result)
+      }
+    })
+    .catch(async e => {
+      try {
+        await handleError(e)
+        emitter.emit('error', e)
+      } catch (newError) {
+        emitter.emit('error', newError)
+      }
+    })
+}
```

Comment:
This would read (slightly) better with async/await:
```js
  try {
    const result = await promise
    if (emitter.listenerCount('data') > 0) {
      emitter.emit('data', { result, id: i })
    } else {
      results.push(result)
    }
  } catch (e) {
    try {
      await handleError(e)
      emitter.emit('error', e)
    } catch (newError) {
      emitter.emit('error', newError)
    }
  }
```

Reply:
02b874d